proposals/dictionary-expressions.md

```diff
@@ -130,41 +186,46 @@ struct Pair<X, Y>
 
 Dictionary<int, string> map1 = [pair1, pair2]; // ?
 
-
 List<Pair<int, string>> pairs = ...;
 Dictionary<int, string> map2 = [.. pairs]; // ?
 ```
 
-### Open question 5
-
-Dictionaries provide two ways of initializing their contents.  A restrictive `.Add`-oriented form that throws when a key is already present in the dictionary, used by collection initializers, and a permissive indexer-oriented form which does not, used by dictionary initializers.  The restrictive form is useful for catching mistakes ("oops, I didn't intend to add the same thing twice!"), but is limiting *especially* in the spread case.  For example:
+Similarly, instead of `KeyValuePair<,>` we could allow *any* type deconstructible to two values.  for example:
 
 ```c#
-Dictionary<string, Option> optionMap = [opt1Name: opt1Default, opt2Name: opt2Default, .. userProvidedOptions];
+record struct Pair<X, Y>(X x, Y y);
+
+Dictionary<int, string> map1 = [pair1, pair2]; // ?
 ```
 
-Or, conversely:
+Resolution: TBD.  Working group recommendation: Allow anything deconstructible into two values (which includes `KeyValuePair`s as well as user types).  It is reasonable to have domain specific types that represent pairs, and to want to then make maps from those.  Making this pleasant, without requiring contortions to do the mapping explicitly would enable powerful and succinct code. 
```

Comment:
the more i thought about this the more i could not come up with a reason to not allow it.  And i *know* users commonly use things like tuples to create interim pairs of things that finally get made into a collection.  Working with KVPs themselves is *unweildy* in many cases (especially due to lack of dedicated language syntax *outside* of what we're adding to collection expressions here).  So tuples are commonly used instead.  being able to say: `Dict<string, int> d = [.. tuplesIveCreated]` is *really* nice.  as are things like `Dict<string, int> d = [.. existingDict.Select(kvp => (kvp.Key, kvp.Value + 1))]`.  Forcing `new KeyValuePair<...>(...)` or `KeyValuePair.Create(...)` here is onerous.

Reply:
Note: we could also take this further when creating collection expressions.  for example:

```c#
List<Point> points = [x1: y1, x2: y2];
```

Namely that not only would we use deconstruction to get the components to create KVPs out of, but that we could also create arbitrary values constructible from two elements.  However, i don't feel that this that necessary.  Especially as one can do: `[new(x1, y1), ...]` which is fairly reasonable.  I think *creation* of arbitrary types (especially as they may allocate) may be a bridge too far.